warehouse/warehouse_test.go

```diff
@@ -0,0 +1,400 @@
+	mockApp.EXPECT().Features().Return(&app.Features{
+		Reporting: report,
+	}).AnyTimes()
+
+	mockBackendConfig := mocksBackendConfig.NewMockBackendConfig(mockCtrl)
+	mockBackendConfig.EXPECT().WaitForConfig(gomock.Any()).DoAndReturn(func(ctx context.Context) error {
+		return nil
+	}).AnyTimes()
+	mockBackendConfig.EXPECT().Identity().Return(nil)
+	mockBackendConfig.EXPECT().Subscribe(gomock.Any(), bcConfig.TopicBackendConfig).DoAndReturn(func(ctx context.Context, topic bcConfig.Topic) pubsub.DataChannel {
+		ch := make(chan pubsub.DataEvent, 1)
+		ch <- pubsub.DataEvent{
+			Data: map[string]bcConfig.ConfigT{
+				workspaceID: {
+					WorkspaceID: workspaceID,
+					Sources: []bcConfig.SourceT{
+						{
+							ID:      sourceID,
+							Enabled: true,
+							Destinations: []bcConfig.DestinationT{
+								{
+									ID:      destinationID,
+									Enabled: true,
+									DestinationDefinition: bcConfig.DestinationDefinitionT{
+										Name: whutils.POSTGRES,
+									},
+								},
+								{
+									ID:      destinationID,
+									Enabled: true,
+									DestinationDefinition: bcConfig.DestinationDefinitionT{
+										Name: whutils.POSTGRES,
+									},
+								},
+							},
+						},
+					},
+				},
+				unsupportedWorkspaceID: {
+					WorkspaceID: unsupportedWorkspaceID,
+					Sources: []bcConfig.SourceT{
+						{
+							ID:      unsupportedSourceID,
+							Enabled: true,
+							Destinations: []bcConfig.DestinationT{
+								{
+									ID:      unsupportedDestinationID,
+									Enabled: true,
+									DestinationDefinition: bcConfig.DestinationDefinitionT{
+										Name: "unknown_destination_type",
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			Topic: string(bcConfig.TopicBackendConfig),
+		}
+		close(ch)
+		return ch
+	}).AnyTimes()
+
+	t.Run("Graceful shutdown", func(t *testing.T) {
+		testCases := []struct {
+			name          string
+			warehouseMode string
+		}{
+			{
+				name:          "master",
+				warehouseMode: config.MasterMode,
+			},
+			{
+				name:          "master and slave",
+				warehouseMode: config.MasterSlaveMode,
+			},
+			{
+				name:          "slave",
+				warehouseMode: config.SlaveMode,
+			},
+		}
+
+		for _, tc := range testCases {
+			tc := tc
+
+			subTestCases := []struct {
+				name        string
+				runningMode string
+			}{
+				{
+					name:        "running mode",
+					runningMode: "",
+				},
+				{
+					name:        "degraded mode",
+					runningMode: degradedMode,
+				},
+			}
+
+			for _, subTC := range subTestCases {
+				subTC := subTC
+
+				t.Run(tc.name+" with "+subTC.name, func(t *testing.T) {
+					pgResource, err := resource.SetupPostgres(pool, t)
+					require.NoError(t, err)
+
+					webPort, err := kithelper.GetFreePort()
+					require.NoError(t, err)
+
+					ctx, stopServer := context.WithCancel(context.Background())
+
+					port, err := strconv.Atoi(pgResource.Port)
+					require.NoError(t, err)
```

Comment:
Can't we skip this step and pass `pgResource.Port` directly in the `c.Set()` below?

Reply:
Since we are getting config.GetInt, I was expecting to set config.Set(<>, <int>). But later realized that [viper uses `cast.ToInt`](https://github.com/spf13/viper/blob/336e9f0a93eec4cec432eadda15d8d0803997d09/viper.go#L987) behind the scenes.

Reply:
Changing now.

Reply:
Thanks for pointing out.